src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs

```diff
@@ -239,30 +247,30 @@ public async Task<GraphNode<RemoteResolveResult>> WalkAsync(LibraryRange library
             return node;
         }
 
-        private Func<LibraryRange, (DependencyResult dependencyResult, LibraryDependency conflictingDependency)> ChainPredicate(Func<LibraryRange, (DependencyResult dependencyResult, LibraryDependency conflictingDependency)> predicate, GraphNode<RemoteResolveResult> node, LibraryDependency dependency)
+        private Func<LibraryRange, (DependencyResult dependencyResult, LibraryDependency conflictingDependency)> ChainPredicate(Func<LibraryRange, (DependencyResult dependencyResult, LibraryDependency conflictingDependency)> predicate,
+                string nodeName,
+                Dictionary<string, LibraryDependency> dependencies,
+                LibraryDependency dependency)
         {
-            var item = node.Item;
-
             return library =>
             {
-                if (StringComparer.OrdinalIgnoreCase.Equals(item.Data.Match.Library.Name, library.Name))
+                if (StringComparer.OrdinalIgnoreCase.Equals(nodeName, library.Name))
                 {
                     return (DependencyResult.Cycle, null);
                 }
 
-                foreach (var d in item.Data.Dependencies)
+                if (dependencies.TryGetValue(library.Name, out var d) &&
```

Comment:
I believe this one will introduce correctness issue. Before logic was at each child node we step back into parent then cross check for cycle or downgrade problem at other child dependency of parent, same time repeatedly go up to another level parent to exhaust of cases. But now it looks we're only looking at 1 branch of code execution not exhaustive as before.

Reply:
Also even though it looks expensive usually C# code is still very fast compare to IO read/write, reflection etc...
So I'm not really sure if we can improve with only simple dictionary, maybe can do large dictionary that can lookup not only library/dependency also has information about which level it was seen, so while we're going top to bottom no need to go up and check with parents. It'll be 1 pass algorithm instead of recursion, but problem is memory might balloon up.

Reply:
> I believe this one will introduce correctness issue. Before logic was at each child node we step back into parent then cross check for cycle or downgrade problem at other child dependency of parent, same time repeatedly go up to another level parent to exhaust of cases. But now it looks we're only looking at 1 branch of code execution not exhaustive as before.
> If we have extensive coverage unit test for this most likely this one cause failure.

I don't think that my modification has changed how it works. It is still going to cross check given node with its parent nodes for cycles and direct dependencies of parent nodes for version eclipsing and downgrades.

Reply:
> So I'm not really sure if we can improve with only simple dictionary, maybe can do large dictionary that can lookup not only library/dependency also has information about which level it was seen, so while we're going top to bottom no need to go up and check with parents. It'll be 1 pass algorithm instead of recursion, but problem is memory might balloon up.

To get rid of recursion to parents I ended up with one dictionary for checking dependencies and one hash set for checking cycles. Unfortunately that has made things even slower than the original linear search. I even tried using immutable data structures to reduce gc pressure (e.g. using `ImmutableHashSet` for cycle detection) but that didn't help either. 
I think that the current approach of replacing linear search with simple dictionary lookup and chaining predicates (so we need to use recursion to parents) is the optimal one. 
I believe that constructing a large dictionary containing dependencies from all levels repeats the work that was done already on previous levels. The time that is saved on dictionary lookups is lost on more expensive dictionary construction and more expensive GC (larger dictionaries mean more memory turnover).

Reply:
> > So I'm not really sure if we can improve with only simple dictionary, maybe can do large dictionary that can lookup not only library/dependency also has information about which level it was seen, so while we're going top to bottom no need to go up and check with parents. It'll be 1 pass algorithm instead of recursion, but problem is memory might balloon up.
> 
> To get rid of recursion to parents I ended up with one dictionary for checking dependencies and one hash set for checking cycles. Unfortunately that has made things even slower than the original linear search. I even tried using immutable data structures to reduce gc pressure (e.g. using `ImmutableHashSet` for cycle detection) but that didn't help either.
> I think that the current approach of replacing linear search with simple dictionary lookup and chaining predicates (so we need to use recursion to parents) is the optimal one.
> I believe that constructing a large dictionary containing dependencies from all levels repeats the work that was done already on previous levels. The time that is saved on dictionary lookups is lost on more expensive dictionary construction and more expensive GC (larger dictionaries mean more memory turnover).

Could be able share your code maybe in another branch or with this PR, I want to see how your dictionary structure, I'm wondering what happened. Many cases runtime complexity can be solved with more memory but sometimes we can reduce both of them. Also include if you have benchmarking code.

Reply:
> Could be able share your code maybe in another branch or with this PR, I want to see how your dictionary structure, I'm wondering what happened. Many cases runtime complexity can be solved with more memory but sometimes we can reduce both of them. Also include if you have benchmarking code.

I've created another branch to avoid clobbering this PR: https://github.com/NuGet/NuGet.Client/compare/dev...marcin-krystianc:marcink-20201118-nochainpredicate

The benchmarking procedure is explained in the description of this PR. In short I was preparing custom dotnet sdks (using release configuration) and then I was using https://github.com/NuGet/NuGet.Client/files/5547424/benchmark_dotnet.ps1.txt to run the benchmark across different solutions.